packages/neptune-extension/src/common/components/upload-modal/UploadForm.tsx

```diff
@@ -0,0 +1,185 @@
+import React from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+
+import { NotebookDTO } from 'generated/leaderboard-client/src/models';
+
+import {
+  PlatformNotebook,
+  PlatformNotebookMetadata,
+} from 'types/platform';
+
+import {
+  uploadNotebook,
+  uploadCheckpoint,
+  getNotebook,
+} from 'common/utils/upload';
+
+import { PROJECT_LOCAL_STORAGE_KEY } from 'common/utils/localStorage';
+
+import Modal from 'common/components/modal/Modal';
+import ProjectInput from 'common/components/input/ProjectInput';
+
+import { getConfigurationState } from 'common/state/configuration/selectors';
+import { getNotebookState } from 'common/state/notebook/selectors'
+import { setNotebook } from 'common/state/notebook/actions'
+
+interface UploadFormProps {
+  platformNotebook: PlatformNotebook
+  onClose: () => void
+}
+
+type UploadMode = 'notebook' | 'checkpoint'
+
+const UploadForm: React.FC<UploadFormProps> = ({
+  platformNotebook,
+  onClose,
+}) => {
+  const [ metadata ] = React.useState(() => platformNotebook.getMetadata());
```

Comment:
In this render, useState nad useSelector lines are mixed and at the first glance it is hard to figure out which props are taken from redux, and which ones are taken from state

Reply:
I barely can see any alternative. I would say even more, in standard code you have even a greater problem, because `mapStateToProps` is hidden on the bottom of the file and you have all props mixed up, those from redux and those from parent. Here at least, you see in the line where you declare a variable.

The only fix I see here is to change `useState` to `useMemo`, because this is not really a state.

Reply:
useMemo would be ok